### PR TITLE
[DataGrid] Performance: small optimizations

### PR DESCRIPTION
**Pull request by romgrk**
Various small optimizations.

Result: small 2-3% improvement
Benchmark: scroll the grid by a constant offset, record the total CPU time

```
branch: next
    N           Min           Max        Median           Avg        Stddev
x  10           163           198           178         181.1     11.483805


branch: perf-small-opts
    N           Min           Max        Median           Avg        Stddev
x   9           145           204           175     176.55556     18.500751
```

### CODE REVIEW

File: packages/x-data-grid/src/components/GridRow.tsx [hunk 1]
```diff
@@ -427,12 +427,12 @@ const GridRow = React.forwardRef<HTMLDivElement, GridRowProps>(function GridRow(
...
         editCellState={editCellState}
         isNotVisible={cellIsNotVisible}
-        {...slotProps?.cell}
         pinnedOffset={pinnedOffset}
         pinnedPosition={pinnedPosition}
         sectionIndex={indexInSection}
         sectionLength={sectionLength}
-        gridHasScrollX={dimensions.hasScrollX}
+        dimensions={dimensions}
+        {...slotProps?.cell}
```
Comment (romgrk) [inline packages/x-data-grid/src/components/GridRow.tsx:435]: Destructuring in the middle creates 3 objects (`_extends({ a: 1}, slotProps.cell, { b: 2 })`).

File: packages/x-data-grid/src/components/cell/GridCell.tsx [hunk 2]
```diff
@@ -326,11 +328,18 @@ const GridCell = React.forwardRef<HTMLDivElement, GridCellProps>((props, ref) =>
...
+      height !== dimensions.rowHeight
+        ? {
+            '--width': `${width}px`,
+            '--height': typeof height === 'number' ? `${height}px` : height,
+            ...styleProp,
+          }
+        : {
+            '--width': `${width}px`,
+            ...styleProp,
+          };
```
Comment (romgrk) [inline packages/x-data-grid/src/components/cell/GridCell.tsx:342]: Avoid setting `--height` unless required. `.setAttribute` is one of our highest costs.

File: packages/x-data-grid/src/hooks/core/pipeProcessing/useGridPipeProcessing.ts [hunk 3]
```diff
@@ -8,12 +8,17 @@ import {
...
 
-interface GridPipeGroupCache {
-  processors: Map<string, GridPipeProcessor<any> | null>;
+type Cache = {
+  [G in GridPipeProcessorGroup]?: GroupCache;
+};
+
+type GroupCache = {
+  processors: Map<string, GridPipeProcessor<any>>;
+  processorsAsArray: GridPipeProcessor<any>[];
```
Comment (romgrk) [inline packages/x-data-grid/src/hooks/core/pipeProcessing/useGridPipeProcessing.ts:17]: Small refactor along with the optimization, the signal-to-noise ratio is high with the long names & prefixed everywhere. This is internal, no need to add the grid prefix.

File: packages/x-data-grid/src/hooks/core/pipeProcessing/useGridPipeProcessing.ts [hunk 4]
```diff
@@ -119,18 +126,16 @@ export const useGridPipeProcessing = (apiRef: React.MutableRefObject<GridPrivate
...
-      }
-
-      return preProcessor(acc, context);
-    }, value);
+    const processors = cache.current[group]!.processorsAsArray;
+    let result = value;
+    for (let i = 0; i < processors.length; i += 1) {
+      result = processors[i](result, context);
+    }
+    return result;
```
Comment (romgrk) [inline packages/x-data-grid/src/hooks/core/pipeProcessing/useGridPipeProcessing.ts:138]: This block is expensive. `Map -> Iterator object -> Array.from() -> array.reduce` is costly when this is run 2x for each cell.

File: packages/x-data-grid/src/utils/createSelector.ts [hunk 5]
```diff
@@ -149,22 +144,24 @@ export const createSelectorMemoized: CreateSelectorFunction = (...args: any) =>
...
+    const cacheArgs = cacheArgsInit ?? new Map();
+    const cacheFn = cacheArgs?.get(args);
 
-    if (cache.get(cacheKey) && cache.get(cacheKey)?.get(args)) {
+    if (cacheArgs && cacheFn) {
       // We pass the cache key because the called selector might have as
       // dependency another selector created with this `createSelector`.
-      return cache.get(cacheKey)?.get(args)(state, cacheKey);
+      return cacheFn(state, cacheKey);
     }
```
Comment (romgrk) [inline packages/x-data-grid/src/utils/createSelector.ts:155]: We're doing `map.get` 4 times but there are only 2 levels of depth. This is called very often.

File: packages/x-data-grid/src/utils/createSelector.ts [hunk 6]
```diff
@@ -173,8 +170,3 @@ export const createSelectorMemoized: CreateSelectorFunction = (...args: any) =>
 
   return selector;
 };
-
-// eslint-disable-next-line @typescript-eslint/naming-convention
-export const unstable_resetCreateSelectorCache = () => {
-  cacheContainer.cache = new WeakMap();
-};
```
Comment (romgrk) [inline packages/x-data-grid/src/utils/createSelector.ts:180]: This seemed to be unused. Ok to remove?

File: packages/x-data-grid-pro/src/components/GridPinnedRows.tsx [hunk 7]
```diff
@@ -38,12 +38,12 @@ export function GridPinnedRows({ position, virtualScroller, ...other }: GridPinn
     ),
   });
 
+  if (rows.length === 0) {
```
Comment (cherniavskii) [inline packages/x-data-grid-pro/src/components/GridPinnedRows.tsx:41]: Can be moved up before the `virtualScroller.getRows` call

File: packages/x-data-grid/src/components/cell/GridCell.tsx (hunk 2 above)
Comment (cherniavskii) [reply packages/x-data-grid/src/components/cell/GridCell.tsx:342]: Do we even need `--height` here? It seems to be always set to `rowHeight`:

https://github.com/mui/mui-x/blob/4e5c0875890542709a0f159d91d4ec9427bf4899/packages/x-data-grid/src/components/GridRow.tsx#L423

File: packages/x-data-grid/src/utils/createSelector.ts (hunk 6 above)
Comment (cherniavskii) [reply packages/x-data-grid/src/utils/createSelector.ts:180]: Huh, we used to call it after each test in karma, but it was removed in https://github.com/mui/mui-x/pull/6043/files#diff-a831d050d6d5aef1f8a23e794a757d885466aa3114f55b2e052b9943ac44981aL13 without explanation (accidentally?).
It was initially added in https://github.com/mui/mui-x/pull/4679

I think we can remove it given we stopped using it 1.5 years ago and didn't even notice 😅

**Comment by romgrk** [comment]
I've addressed the comments. The failing test is a flaky datepicker test. Approved?

**Review by cherniavskii — APPROVED** [review]
🎉

**Pull request merged**